File: 2025/typescript/day04.ts
import { type DayModule, type SolutionFactory } from './types'
import * as utils from './utils'

export const part1: Part = input => () => {
    const grid = input.map(line => line!.split(''))
    return findRemovableRolls(grid).length
}

export const part2: Part = input => () => {
    const grid = input.map(line => line!.split(''))

    let count = 0
    let rolls = findRemovableRolls(grid)
    while (rolls.length > 0) {
        count += rolls.length

        const nextCandidates: Map<string, [number, number]> = new Map()
        for (const pos of rolls) {
            const [x, y] = pos
            grid[y]![x] = '.'
            for (const pos of adjacentPositions(x, y, grid)) {
                const [nx, ny] = pos
                if (grid[ny]![nx] !== '@') continue
                nextCandidates.set(`${nx},${ny}`, pos)
            }
        }

        rolls = findRemovableRolls(grid, Array.from(nextCandidates.values()))
    }

    return count
}

function findRemovableRolls(
    grid: string[][],
    process?: [number, number][],
): [number, number][] {
    const width = grid[0]!.length
    const height = grid.length
    const rolls: Map<string, [number, number]> = new Map()
    const canRemove = (x: number, y: number) => {
        if (grid[y]![x] !== '@') return false
        let adjacentRollsCount = 0
        for (const [nx, ny] of adjacentPositions(x, y, grid)) {
            if (grid[ny]![nx] !== '@') continue
            if (++adjacentRollsCount > 3) return false
        }
        return true
    }
    const addIfRemovable = (pos: [number, number]) => {
        const [x, y] = pos
        if (canRemove(x, y)) rolls.set(`${x},${y}`, pos)
    }

    if (process) {
        for (const pos of process) addIfRemovable(pos)
    } else {
        for (let y = 0; y < height; y++)
            for (let x = 0; x < width; x++) addIfRemovable([x, y])
    }
    return Array.from(rolls.values())
}

function* adjacentPositions(
    x: number,
    y: number,
    grid: string[][],
): IterableIterator<[number, number]> {
    const width = grid[0]!.length
    const height = grid.length
    const deltas = [-1, 0, 1]
    for (const dx of deltas) {
        for (const dy of deltas) {
            if (dx === 0 && dy === 0) continue
            const nx = x + dx
            const ny = y + dy
            if (ny >= 0 && ny < height && nx >= 0 && nx < width) {
                yield [nx, ny]
            }
        }
    }
}

export const day = import.meta.file.match(/day(\d+)/)![1]!
export const input = await utils.readInputLines(day)
part1.solution = 1523
part2.solution = 9290

export const main = import.meta.main
if (main) {
    const module: Module = await import(import.meta.path)

    const testInput = utils.as_lines([
        `
        ..@@.@@@@.
        @@@.@.@.@@
        @@@@@.@.@@
        @.@@@@..@.
        @@.@@@@.@@
        .@@@@@@@.@
        .@.@.@.@@@
        @.@@@.@@@@
        .@@@@@@@@.
        @.@.@@@.@.
        `,
    ])

    await utils.tests(
        () =>
            utils.test_all(
                ['Test part 1', 13, part1(testInput[0]!)],
                ['Test part 2', 43, part2(testInput[0]!)],
            ),
        () => utils.test_day(module),
    )
}

type In = typeof input
type Out = number
type Module = DayModule<Out, In>
type Part = SolutionFactory<Out, In>

File: 2025/zig/src/day04.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const aoc = @import("aoc_utils");

const Field = enum(u8) { empty = '.', roll = '@' };
const Grid = aoc.Grid(Field, u8);

pub fn part1(input: []const u8, gpa: Allocator) !Day.Result1 {
    const grid = try Grid.init(input, gpa);
    defer grid.deinit(gpa);

    const rollsBuffer = try gpa.alloc(Grid.Pos, grid.buf.len);
    defer gpa.free(rollsBuffer);

    const rolls = try findRemovableRolls(&grid, null, rollsBuffer);
    return @intCast(rolls.len);
}

pub fn part2(input: []const u8, gpa: Allocator) !Day.Result2 {
    var grid = try Grid.init(input, gpa);
    defer grid.deinit(gpa);

    const rollsBuffer = try gpa.alloc(Grid.Pos, grid.buf.len);
    defer gpa.free(rollsBuffer);

    var count: u16 = 0;
    var rolls = try findRemovableRolls(&grid, null, rollsBuffer);

    var nextCandidates = std.AutoArrayHashMapUnmanaged(Grid.Pos, void).empty;
    defer nextCandidates.deinit(gpa);
    while (rolls.len > 0) {
        count += @intCast(rolls.len);

        nextCandidates.clearRetainingCapacity();
        for (rolls) |pos| grid.setAt(pos, .empty);
        for (rolls) |pos| {
            for (dirs) |dir| {
                if (dir.nextOrNull(grid, pos)) |next| {
                    if (grid.at(next) == .roll) {
                        try nextCandidates.put(gpa, next, {});
                    }
                }
            }
        }

        rolls = try findRemovableRolls(&grid, nextCandidates.keys(), rollsBuffer);
    }

    return count;
}

fn findRemovableRolls(grid: *const Grid, process: ?[]Grid.Pos, buffer: []Grid.Pos) ![]Grid.Pos {
    var rolls = std.ArrayList(Grid.Pos).initBuffer(buffer);

    if (process) |positions| {
        for (positions) |pos| {
            if (canRemoveRoll(grid, pos)) try rolls.appendBounded(pos);
        }
    } else {
        for (grid.buf, 0..) |field, index| {
            if (field != .roll) continue;
            const pos = grid.indexToPos(index);
            if (canRemoveRoll(grid, pos)) try rolls.appendBounded(pos);
        }
    }

    return rolls.items;
}

const dirs = std.enums.values(aoc.Direction);
fn canRemoveRoll(grid: *const Grid, pos: Grid.Pos) bool {
    std.debug.assert(grid.at(pos) == .roll);
    var adjacentRollsCount: u8 = 0;
    for (dirs) |dir| {
        if (dir.nextOrNull(grid.*, pos)) |next| {
            if (grid.at(next) != .roll) continue;
            adjacentRollsCount += 1;
            if (adjacentRollsCount > 3) return false;
        }
    }
    return true;
}

pub const Day = aoc.DayInfo("04", u16, u16, 1523, 9290, @This(), &.{.{
    .expected1 = 13,
    .expected2 = 43,
    .input =
    \\..@@.@@@@.
    \\@@@.@.@.@@
    \\@@@@@.@.@@
    \\@.@@@@..@.
    \\@@.@@@@.@@
    \\.@@@@@@@.@
    \\.@.@.@.@@@
    \\@.@@@.@@@@
    \\.@@@@@@@@.
    \\@.@.@@@.@.
    ,
}});

test "samples 1" {
    try Day.testPart1Samples();
}
test "samples 2" {
    try Day.testPart2Samples();
}
test "part 1" {
    try Day.testPart1();
}
test "part 2" {
    try Day.testPart2();
}
